Dynamic fees V1 (#1)
mobilecoinfoundation/rfcs: text/0001-dynamic-fees-v1.md

Choosing to simply accept that if any node has accepted a proposed transaction at a lower fee, that fee is (or would have been) valid for all nodes is an option, but this does make the ability to enforce fees dependent on the union of all SGX functionality, which is a change from our ad-hoc norms around the use of SGX as a privacy-enhancement technology, rather than a security-critical technology.

The second option is to only check fees lazily, which will produce the scenario where a client could submit a transaction, and have it appear to "time out" later, because it was accepted into the queue at a low fee, then later judged to have an insufficient fee.

The "restart for each change" system will have similar properties for users as an emergency restart to introduce new fees---when a node is restarted without following the upgrade procedure, any pending transactions it has are lost---but is dramatically simpler to implement, and does not require special thought about validity periods of a particular transactions' fee-paid status.

> ## Alt: Allow Divergent Fees

Reviewer: Divergent fee makes a lot of sense to me. For example I may be willing to conduct transaction processing when spot instances come up for sale on aws/azure/etc but not when that is not the case. This allows me to operate at 10% of the cost. I can leave a "pilot light" up at higher costs and then spin up unlimited clusters of more expensive software when that happens.
This makes the network almost impossible to ddos you'd have to be able to take down aws, azure, and gcloud simultaneously, and at that point why waste it on mob

Author: My instincts prefer divergent fees as well, but there's a bunch of gotchas that we'd have to talk our way through in order to get it to a place that feels comfortable---IMO it's a great candidate for the *next* "dynamic fees" RFC, where a fuller discussion is possible without the "OMG, tx costs $0.50" overhang :-)

Reviewer: So I can speak as a global expert when I say the following: Unless you standardize deployment in such a way that all the node operators have a couple of unneeded SPOF, everybody's infra is gonna run at a different cost, and whats more, at a certain repayment level, we can easily handle more load, its pretty much a demand economy in cloud nowadays.

(I will personally audit and make recommendations to node operators to run cheaper if you @ me on socials)

Reviewer: If a tx costs 50c to run, yall arent running all of the nodes in an efficient way. The aws minimum cost for an entire hour of time on a server that can run a (virtualized) SGX is 0.192000 ... are we really running one tx every 2.5 hours per node... seems unlikely.

Author: The "$0.50" refers to 0.01MOB fee at $50/MOB. Whatever the actual costs to validating node operators, it costs *users* $0.50 (well, $0.325 as of this writing) to conduct a transaction.

- Feature Name: Dynamic Fees V1 (Configurable Fees)
- Start Date: (2021-04-20)
- RFC PR: [mobilecoinfoundation/rfcs#1](https://github.com/mobilecoinfoundation/rfcs/pull/1)
- MobileCoin Epic: MCC-2301

# Summary
[summary]: #summary

Decrease the minimum fee to 0.0004 MOB (400uMOB, which is 0.02 USD/tx at a 50 USDMOB exchange rate) per board instruction, allow node operators to override the hard-coded minimum fee, and prevent divergent fees by including fee agreement in the node-to-node attestation / key exchange protocol.

# Motivation
[motivation]: #motivation

MobileCoin launched with a hard-coded minimum fee of 0.01 MOB. When this constant was set, the value of MobileCoin was assumed to be slightly below 1 USD per coin, resulting in a per-transaction fee of under 0.01 USD (1c). In the intervening 6 months, the value per-coin increased to 50 USD, resulting in a per-transaction fee of 0.50 USD.

This fee is too high to reasonably support the cannonical use-case---purchasing a cup of coffee---and we would like to lower the fee to $0.04 to bring it back in line with that use-case.

> However, this creates up a different problem. If the consensus system is roughly capable of completing 50tps (transactions per second), then the cost of conducting a denial of service attack is directly related to the size of the fees. At 0.50 USD/tx, consuming the entire consensus power of the network would cost over 2 MUSD/day.

Reviewer 2: ```suggestion
However, this creates a different problem. If the consensus system is roughly capable of completing 50tps (transactions per second), then the cost of conducting a denial of service attack is directly related to the size of the fees. At 0.50 USD/tx, consuming the entire consensus power of the network would cost over 2M USD/day.
```

> At the desired 0.04 USD/tx, however, the cost of conducing a denial of service attack would be less than 175 kUSD/day. This may seem like a lot of money, but recall that at peak trading on FTX, 80 MMOB or 4 GUSD changed hands, which gives some sense of scale to those costs (i.e. they are too low).

Reviewer 2: ```suggestion
At the desired 0.04 USD/tx, however, the cost of conducing a denial of service attack would be less than 175k USD/day. This may seem like a lot of money, but recall that at peak trading on FTX, 80M MOB or 4G USD changed hands, which gives some sense of scale to those costs (i.e. they are too low).
```

Reviewer 2: Hmm - I see why you indicated the SI unit modifier on the currency as opposed to the number - I was trying to avoid the confusion of "MUSD" = "MobileUSD" which could be a future stablecoin ticker.

Author: Right, I think MoUSD might be an elegant solution for the naming of future stuff ;-)

The second issue is that changing the hard-coded fee requires going through the full release cycle: testnet deployment and testing, mainnet enclave signing, client updates to trust the new enclave, a grace period for client updates, then finally a coordinated mainnet deployment by operators.

In our current environment, it takes at least a month to change the fee, and the fees are simultaneously too high and too low. This must change.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> When a node operator starts the consensus service, it may optionally provide a configuration option or command-line argument overriding fee. The fee will be appended (as a string) to the existing to the key exchange "prologue" (a series of bytes that both sides agree upon in advance) when performing node-to-node attestation.

Reviewer 2: ```suggestion
When a node operator starts the consensus service, it may optionally provide a configuration option or command-line argument overriding fee. The fee will be appended (as a string) to the existing key exchange "prologue" (a series of bytes that both sides agree upon in advance) when performing node-to-node attestation.
```

> If two nodes with different fees attempt to communicate, the result will be an `AeadError` during the handshake phase.

Reviewer 2: Could this error be more descriptive? e.g. `AeadError(FeeMismatch)`

Author: Unfortunately no: the noise prologue is just a byte string that each side independently hashes together with the key exchange results to construct the symmetric key, and this proposal is saying "let's extend that to include the fee, as well as the [existing] responder ID", so now, if you get an `AeadError` when you are doing an attestation, it means "you and the node you are connecting do disagree about their responder ID (or there is a MiTM attack that has replaced a pubkey somewhere)", now would be updated to be: "disagree about their responder ID or minimum fee".

> The currently-configured fee should be made available to consensus clients directly by extending the `BlockchainAPI.GetLastBlockInfo` response. This gRPC call is already polled by thick clients to orient themselves with relation to the state of the network.

Reviewer 2: 👍

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The `consensus_service` application will be extended to add a new `--minimum-fee` command-line argument and `MC_MINIMUM_FEE` environment variable. These values will be merged (environment-variable, then command-line option). This value will contain a string indicating the quantity and SI-scale of the minimum fee. Fees with out units will be considered to be denominated in picoMOB. Therefore, all these values are equivilent fees:

Reviewer 2: "merged" meaning command-line option supersedes env var?

Author: Yes, I'll update.

The given minimum fee value, if it is provided, will be injected into the enclave at startup and stored for future use when validating inbound transactions. As is the case today, fees will be validated initially, on transaction submission.

This fee value will be injected into the attested key exchange by the enclave, by constructing a new UTF-8 string of the format: `"%s-%u"` using the existing `ResponderId`, and the u64 picoMOB-denominated representation of the new fee. This value is injected into the hashing scheme built onto the key exchange protocols in use.

Because of this, two nodes with different fees will produce different encryption keys, which will lead to a failed decyption during the key exchange setup (specifically, the initiator will receive an `AeadError` when attempting to decrypt the initial message received from the responder).

# Drawbacks
[drawbacks]: #drawbacks

> The primary drawback to this approach is that it requires a restart of a quorum of the network in order to adjust to a new fee. In the intended common-case (low fees generally, higher fees in response to an active denial-of-service), this will entail restarting the network twice, and when a node is restarted, any pending transactions which did not make it into the current SCP slot will be discarded.

Reviewer 2: Worth mentioning that the restart protocols include letting the pending values drain while the nodes are set to not accept new transactions?

> 

Reviewer 2: If it's not possible to be more descriptive in the `AeadError`, which it seems like it might not be, then I would add the challenge of debugging the attestation failure as a drawback.

Author: Fair point.

The next alternative was using a "fee schedule", or "fee calendar"---that is, rather than having a single fee which is constructed at startup time, use a configuration file which provides a series of "fee ranges" `(fee, startBlock, endBlock)` that are hashed into the key exchange.

This alternative scheme does have the advantage of allowing a reconfiguration to "script" the denial-of-service response, and it's subsequent unwind, in a single restart. The devil, however, is in the details: how should pending and future transactions be handled?

This is actually quite tricky to get right: obviously you want to immediately reject any transactions which do not meet the currently configured fee requirements, but it's entirely possible that your node is already queuing transactions when this comes in.

In this scenario, a client submits a transaction, and it goes into a queue behind hundreds of other transactions. The fee is "correct" for the current block, but the transaction will not actually be entertained for acceptance until after the fee has been raised, and the fee is no longer valid.

> Chosing to simply accept that if any node has accepted a proposed transaction at a lower fee, that fee is (or would have been) valid for all nodes is an option, but this does make the ability to enforce fees dependent on the union of all SGX functionality, which is a change from our ad-hoc norms around the use of SGX as a privacy-enhancement technology, rather than a security-critical technology.

Reviewer 2: ```suggestion
Choosing to simply accept that if any node has accepted a proposed transaction at a lower fee, that fee is (or would have been) valid for all nodes is an option, but this does make the ability to enforce fees dependent on the union of all SGX functionality, which is a change from our ad-hoc norms around the use of SGX as a privacy-enhancement technology, rather than a security-critical technology.
```

The second option is to only check fees lazily, which will produce the scenario where a client could submit a transaction, and have it appear to "time out" later, because it was accepted into the queue at a low fee, then later judged to have an insufficient fee.

> The "restart for each change" system will have similar properties for users---when a node is restarted, any pending transactions it has are lost---but is dramatically simpler to implement, and does not require special thought about validity periods of a particular transactions' fee-paid status.

Reviewer 2: Another place to potentially clarify that the pending transactions are only lost in an emergency restart scenario, not if operators are able to follow restart protocols.

The primary rationale, then, is that this is "good enough" for right now, and we will follow this up with a "dynamic fees v2" that settles all questions properly under easier time constraints.

# Prior art
[prior-art]: #prior-art

> Unfortunately, due to the limited scope of this PR, prior art is going to be difficult to discuss. There is obviously a much larger conversation to be had about Etherium gas, Bitcoin transaction fees, etc., and how MobileCoin fees differ from them, but that honestly belongs in a future RFC. This is about doing the minimum viable to get us to a baseline fees, without compromising our ability to respond to denial of service attacks.

Reviewer 2: ```suggestion
Unfortunately, due to the limited scope of this PR, prior art is going to be difficult to discuss. There is obviously a much larger conversation to be had about Etherium gas, Bitcoin transaction fees, etc., and how MobileCoin fees differ from them, but that honestly belongs in a future RFC. This is about doing the minimum viable to get us to a baseline fee-adjustment strategy, without compromising our ability to respond to denial of service attacks.
```

This is actually quite tricky to get right: obviously you want to immediately reject any transactions which do not meet the currently configured fee requirements, but it's entirely possible that your node is already queuing transactions when this comes in.

In this scenario, a client submits a transaction, and it goes into a queue behind hundreds of other transactions. The fee is "correct" for the current block, but the transaction will not actually be entertained for acceptance until after the fee has been raised, and the fee is no longer valid.

Choosing to simply accept that if any node has accepted a proposed transaction at a lower fee, that fee is (or would have been) valid for all nodes is an option, but this does make the ability to enforce fees dependent on the union of all SGX functionality, which is a change from our ad-hoc norms around the use of SGX as a privacy-enhancement technology, rather than a security-critical technology.

> The second option is to only check fees lazily, which will produce the scenario where a client could submit a transaction, and have it appear to "time out" later, because it was accepted into the queue at a low fee, then later judged to have an insufficient fee.

Reviewer 3: In practice:
1. there should be an expectation in the ecosystem about how much the 'lazy' minimum fee can vary over short timespans
2. the default fee used by clients can be some multiple of the minimum fee, above the potential variance over e.g. 100 blocks (current tombstone limit)

Reviewer 3: actually, both of those points apply to any solution, lazy or otherwise
